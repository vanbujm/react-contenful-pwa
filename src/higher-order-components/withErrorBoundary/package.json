{
  "name": "withErrorBoundary",
  "main": "withErrorBoundary.js"
}
