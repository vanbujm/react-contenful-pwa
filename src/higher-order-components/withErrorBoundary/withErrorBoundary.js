import React from 'react';
import { lifecycle, setDisplayName, wrapDisplayName, compose } from 'recompose';

export const withErrorBoundary = (
  FallbackComponent,
  errorHandler
) => BaseComponent => {
  const lifeCycleSpec = {
    componentDidCatch(error, info) {
      if (errorHandler !== undefined) {
        errorHandler(error, info);
      }
      this.setState({ hasError: true });
    }
  };

  const ErrorBoundary = props => {
    const { hasError } = props;
    const ActualFallbackComponent = FallbackComponent ? (
      <FallbackComponent {...props} />
    ) : (
      <h1>Yikes, something went wrong.</h1>
    );

    if (hasError) return ActualFallbackComponent;
    return <BaseComponent {...props} />;
  };

  if (process.env.NODE_ENV !== 'production') {
    return compose(
      setDisplayName(wrapDisplayName(BaseComponent, 'withErrorBoundary')),
      lifecycle(lifeCycleSpec)
    )(ErrorBoundary);
  }

  return lifecycle(lifeCycleSpec)(ErrorBoundary);
};

export default withErrorBoundary;
