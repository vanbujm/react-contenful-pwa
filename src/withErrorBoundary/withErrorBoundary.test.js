import React from 'react';
import { mount } from 'enzyme';

import { withErrorBoundary } from './withErrorBoundary';

const actualErrorLogger = console.error;

const ComponentThatBreaks = () => {
  throw new Error('Boom');
};

beforeAll(() => {
  // error boundary catches the error but logs it in console.error,
  //  we don't want to see the log every time the test is run
  console.error = () => null;
});

afterAll(() => {
  // Put the real console.error method back for other tests
  console.error = actualErrorLogger;
});

it('Catches errors', () => {
  const ComponentThatCatchesErrors = withErrorBoundary()(ComponentThatBreaks);
  expect(mount(<ComponentThatCatchesErrors />)).toMatchSnapshot();
});

it('Excepts a different error component', () => {
  const ComponentThatCatchesErrors = withErrorBoundary(() => (
    <h1>Wow, an error...</h1>
  ))(ComponentThatBreaks);
  expect(mount(<ComponentThatCatchesErrors />)).toMatchSnapshot();
});

it('Accepts an error handler', () => {
  let anError = null;
  const ComponentThatCatchesErrors = withErrorBoundary(
    null,
    error => (anError = error)
  )(ComponentThatBreaks);
  mount(<ComponentThatCatchesErrors />);
  expect(anError).not.toBeNull();
});
