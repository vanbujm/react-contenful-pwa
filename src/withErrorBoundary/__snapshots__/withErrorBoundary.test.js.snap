// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catches errors 1`] = `
<withErrorBoundary(ComponentThatBreaks)>
  <ErrorBoundary
    hasError={true}
  >
    <h1>
      Yikes, something went wrong.
    </h1>
  </ErrorBoundary>
</withErrorBoundary(ComponentThatBreaks)>
`;

exports[`Excepts a different error component 1`] = `
<withErrorBoundary(ComponentThatBreaks)>
  <ErrorBoundary
    hasError={true}
  >
    <Component
      hasError={true}
    >
      <h1>
        Wow, an error...
      </h1>
    </Component>
  </ErrorBoundary>
</withErrorBoundary(ComponentThatBreaks)>
`;
